### PR TITLE
[READY]Custom reagent icecream + tweaks

### PR DESCRIPTION
**Pull request by Krysonism**
milkies

## About The Pull Request

This pull requests lets the ice cream vat dispense custom ice creams flavoured with a reagent of your choice from an internal beaker. The ice cream takes the name and reagent colour of the master reagent.

You can also easily refill the ice cream vats ingredient storage by inserting a beaker and clicking the refill from beaker button.

I also plan on adjusting the colour of a few reagent that i think deserve good looking ice creams but do not have the colour for it.

## Why It's Good For The Game

The ice cream vat is currently really lame, the fun comes from making strange and interesting flavours, a feature that is absent from /tg/. No more!

The ice creams in this screenshot are flavoured with earthsblood, crank, romerol, krokodil, turbo and happiness, but the possibilities are endless, you can be bold and incorporate some mayonnaise or just play it safe with caramel.

## Changelog
:cl:
add: NT has made breakthroughs in ice cream science, ice creams can now be flavoured with any reagent!
tweak: The ice cream vat now accepts beakers.
/:cl:

### CODE REVIEW

**Comment by TerraGS** [comment]
Liquid Gibs ice cream

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm [hunk 1]
```diff
@@ -211,6 +252,21 @@
 		if ("mob")
 			desc = "A suspicious [cone_type] cone filled with bright red ice cream. That's probably not strawberry..."
 			reagents.add_reagent(/datum/reagent/liquidgibs, 2)
+		if ("custom")
+			if(R && R.total_volume >= 2)
```
Comment (qwertyquerty) [inline code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: `R?.total_volume >= 2`

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm [hunk 2]
```diff
@@ -76,15 +82,22 @@
...
+		dat += "<b>BEAKER CONTENT</b><br><div class='statusDisplay'>"
+		for(var/datum/reagent/R in beaker.reagents.reagent_list)
+			dat += "[R.name]: [R.volume]u<br>"
+		dat += "<a href='?src=[REF(src)];refill=1'><b>Refill from beaker</b></a></div>"
+	dat += "<br>"
 	dat += "<b>VAT CONTENT</b><br>"
 	for(var/datum/reagent/R in reagents.reagent_list)
 		dat += "[R.name]: [R.volume]"
-		dat += "<A href='?src=[REF(src)];disposeI=[R.type]'>Purge</A><BR>"
+		dat += "<A href='?src=[REF(src)];disposeI=[R.type]'>Purge</a><BR>"
```
Comment (qwertyquerty) [inline code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: You randomly changed `</A>` to `</a>` here which doesn't match the previous tag.

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm [hunk 3]
```diff
@@ -222,10 +278,28 @@
 		new /obj/item/stack/sheet/metal(loc, 4)
 	qdel(src)
 
+/obj/machinery/icecream_vat/AltClick(mob/living/user)
+	if(!istype(user) || !user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+		return
+	replace_beaker(user)
+	return
```
Comment (qwertyquerty) [inline code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: Not sure the purpose of this return

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm [hunk 4]
```diff
@@ -18,6 +19,7 @@
 	var/list/product_types = list()
 	var/dispense_flavour = ICECREAM_VANILLA
 	var/flavour_name = "vanilla"
+	var/obj/item/reagent_containers/beaker = null
```
Comment (qwertyquerty) [inline code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: I believe you are allowing any reagent container to work not just beakers. Keep in mind a lot of things are classified as reagent containers.

Comment (GuyonBroadway) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: This means you can make pizza flavoured icecream, you tell me that isn't a hilarious notion.

**Comment by GuyonBroadway** [comment]
>That picture is named milkies.png

On a more serious note if you are going to borrow the goonstation feature of reagent icecream consider also borrowing the goonstation idea of the bottles to cheap drinks/condiments that contain a random assortment reagents that range from chilli juice to ants.

**Comment by necromanceranne** [comment]
Meth icecream for a meth icecream cake with meth icing was how I killed a heavily armed traitor in hand to hand combat on goon once. He kept asking how I was still alive froma combination of baton smacks and a bullet lodged in me and I screamed, in my horrible swedish accent, METH CAKE.

Anyway I look forward to combining with botany to produce more meth ice cream cakes.

**Comment by Krysonism** [comment]
> 
> 
> Meth icecream for a meth icecream cake with meth icing was how I killed a heavily armed traitor in hand to hand combat on goon once. He kept asking how I was still alive froma combination of baton smacks and a bullet lodged in me and I screamed, in my horrible swedish accent, METH CAKE.
> 
> Anyway I look forward to combining with botany to produce more meth ice cream cakes on tg as well. I've been wanting to see this feature for ages! It actually gives a good reason for chefs to reach out to their coworkers to produce cool food. Like the bartender, botany and chemistry.

Good point, i will try to make sure the ice cream transfers the correct colour to whichever custom food incorporates it.

By the way, currently the ice cream vat only transfers 2u of the reagent to the icecream to be consistent with the other flavour types, should we increase this somehow to 3 or 4, or should we just let the chef handle it by`manual syringe injection?

**Comment by Krysonism** [comment]
> 
> 
> Liquid Gibs ice cream

I will probably make liquid gibs a deeper red colour since it makes a muddy orange icecream right now.

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 1 above)
Comment (Krysonism) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: I will fix this today

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 2 above)
Comment (Krysonism) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: I must have done it trying to get dispose to  work. Any idea why the dispose button is broken?

It was broken before i started messing with the vat...

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 3 above)
Comment (Krysonism) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: I will purge it

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 4 above)
Comment (Krysonism) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: I don't think the machine will accept any non-open containers such as food, pills or patches.

**Comment by GuyonBroadway** [comment]
Clf3 ice cream should be called "spice-cream"

**Comment by OverlyGeneric** [comment]
we goon now boys. 

now we need randomized maintenance ice cream

**Comment by necromanceranne** [comment]
Maybe have it be a setting in the ice cream vat for how many units of a reagent is used? 2u is likely not to do much with the fast metabolism reagents/toxins but that might be a good thing.

**Comment by Krysonism** [comment]
> 
> 
> Maybe have it be a setting in the ice cream vat for how many units of a reagent is used? 2u is likely not to do much with the fast metabolism reagents/toxins but that might be a good thing.

I have increased it to 3u. 

If the chef desires to poison his ice cream he can inject 10u additional reagents with a syringe.

I might do the button ID later.

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 4 above)
Comment (qwertyquerty) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: Ah I see now

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 2 above)
Comment (qwertyquerty) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: Took a look and I'm not sure

**Comment by Krysonism** [comment]
Has anyone an idea why the dispose button is not working?

**Comment by Krysonism** [comment]
I will be leaving for vacation in Germany for a week so i might not be able to make any edits until next week.

**Comment by WaterStBlues** [comment]
Ice icecream

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 4 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: the last `/` is the var name so the `beaker` var is typed as `/obj/item/reagent_containers`

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 2 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]: it still doesn't work after text2path?

**Comment by redmoogle** [comment]
I want this

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 2 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]:
```suggestion
		dat += "<A href='?src=[REF(src)];disposeI=[R.type]'>Purge</A><BR>"
```

File: code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm (hunk 3 above)
Comment (ExcessiveUseOfCobblestone) [reply code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm]:
```suggestion
```

**Comment by Krysonism** [comment]
I am back in Sweden and ready address any suggested changes.

**Pull request merged**